### ISSUE TITLE
Need help: NUTS does't converge

### ISSUE THREAD

**Issue by fangzefunny** [pyro-ppl/numpyro#1241]
I observed my numpyro implementation gives very different results from a JAGS model (working really well using 1000 warmups). The main issue is the chain does not converge regardless of warmup numbers. Here are some notes:

* I implement the numpyro model wrong。 
* Maybe I need to initiate the chain but I did not.
* Sampling looks a bit weird when I implement an exponential function or a power law. Right before the end, the acc.prob usually drops from .8 to .3. 
* I try to increase the warmups samples to 50,000, still does not converge

My code:

```python 

import numpy as np
import numpyro as npyro
import arviz as az
import jax.numpy as jnp

from jax import random 
from scipy.stats import norm, binom 
from numpyro.infer import MCMC, NUTS, Predictive

def hMF( obs, eps=.001):

    ## Populational level distributions
    mu_alpha  = npyro.sample( 'mu_alpha', dist.Uniform( 0, 1))
    tau_alpha = npyro.sample( 'tau_alpha', dist.Gamma( eps, eps))
    mu_a      = npyro.sample( 'mu_a', dist.Uniform( 0, 1))
    tau_a     = npyro.sample( 'tau_a', dist.Gamma( eps, eps))
    mu_b      = npyro.sample( 'mu_b', dist.Uniform( 0, 1))
    tau_b     = npyro.sample( 'tau_b', dist.Gamma( eps, eps))

    ## Individual distributions
    alpha = npyro.sample( 'alpha', dist.TruncatedDistribution(
                    dist.Normal( jnp.broadcast_to( mu_alpha, [obs['n_subj'], 1]), tau_alpha), low=0, high=1))
    a     = npyro.sample( 'a', dist.TruncatedDistribution( 
                    dist.Normal( jnp.broadcast_to( mu_a, [obs['n_subj'], 1]), tau_a), low=0, high=1))
    b     = npyro.sample( 'b', dist.TruncatedDistribution( 
                    dist.Normal( jnp.broadcast_to( mu_b, [obs['n_subj'], 1]), tau_b), low=0, high=1))

    ## Each time steps
    sub_id = np.arange( obs['n_subj'])
    t_lags = np.arange( obs['T']).reshape([1,-1])
    theta = a[sub_id] + (1 - a[sub_id]) * b[sub_id] *\
        jnp.exp(-alpha[sub_id] * t_lags)
    
    return npyro.sample( 'recall', dist.Binomial( obs['n_items'], theta
                                ), obs=obs['recalls'])

def sim_hMF( seed=1234):

    ## Generate data
    rng = np.random.RandomState( seed)
    tlags = [ 0, 1, 5, 10, 20, 50]
    n_subj, T, n_items = 4, len(tlags), 20
    recalls = np.zeros( [ n_subj, T]) + np.nan 
    for sub_i in range(n_subj):
        a = .2*rng.rand()
        b = .1*rng.rand() + .9
        alpha = .3*rng.rand() + .1
        print( f'Sub {sub_i}: a={a:.2f}, b={b:.2f}, alpha={alpha:.2f}')
        for j, t in enumerate(tlags):
            p = a + (1-a) * b * np.exp(-alpha*t)
            recalls[ sub_i, j] = binom( n=n_items, p=p).rvs()
    obs = { 'n_subj': n_subj, 'n_items': 20, 'T': T,
            'recalls': recalls}

    ## Fix the random seed
    rng_key = random.PRNGKey(seed)

    ## Sampling 
    kernel = NUTS( model)
    posterior = MCMC( kernel, num_chains=1, 
                              num_samples=5000, 
                              num_warmup=10000,)
    posterior.run( rng_key, obs)
    samples = posterior.get_samples()
    
    ## Summarize the sampling results
    print( posterior.print_summary())
    az.plot_trace( az.from_numpyro( posterior), compact=True)
```

My az figures

![image](https://user-images.githubusercontent.com/24639671/143691023-ee5674be-685e-43f8-bd72-0145232e4351.png)

**Comment by martinjankowiak** [comment]
- your priors over the `tau_` variables have very high variance is that what you want? if so you may want to initialize those variables to 1 or the like
- it doesn't appear that `theta` is a probability (?). when using `Binomial` you should explicitly specify if you're passing in `probs` or `logits`

**Comment by fangzefunny** [comment]
Thanks for your quick reply. I will consider playing around with the tau parameters but the benchmark JAGS model uses the same distribution but doing really well. So I prefer to first stick to this setting. 

Now I specify the ```dist.Binomial( total_count=obs['n_items'], probs=theta)```, still the sampling looks unexpected. 
![image](https://user-images.githubusercontent.com/24639671/143722297-5d0c437f-aad5-4caf-b9f5-5084286ee3e1.png)

**Comment by fehiepsi** [comment]
> implement the numpyro model wrong

You can draw samples from prior predictive (using Predictive without posterior_samples and with `obs=None` in the likelihood) to compare with your jags prior predictive. This way you can be more confident that you have an expected model. Something to check:
+ Some distributions might have different parameterizations.
+ Add some assertion for the shapes
+ Use [render_model](http://num.pyro.ai/en/stable/utilities.html#numpyro.contrib.render.render_model) and [format_shapes](http://num.pyro.ai/en/stable/utilities.html#numpyro.util.format_shapes) to double-check that you have the expected model. For this, you need to add `plate` statements or `.to_event()` at `alpha`, `a`, `b` sites, e.g.
```
with plate("n_subj", obs['n_subj']):
    alpha = npyro.sample( 'alpha', dist.TruncatedDistribution(
                    dist.Normal(mu_alpha, tau_alpha), low=0, high=1))
    alpha = alpha[..., None]
```
or
```
    alpha = npyro.sample( 'alpha', dist.TruncatedDistribution(
                    dist.Normal(mu_alpha, tau_alpha), low=0, high=1).expand([obs['n_subj'], 1]).to_event(2))
```

**Comment by fangzefunny** [comment]
> * Some distributions might have different parameterizations.
> * Add some assertion for the shapes
> * Use [render_model](http://num.pyro.ai/en/stable/utilities.html#numpyro.contrib.render.render_model) and [format_shapes](http://num.pyro.ai/en/stable/utilities.html#numpyro.util.format_shapes) to double-check that you have the expected model. For this, you need to add `plate` statements or `.to_event()` at `alpha`, `a`, `b` sites, e.g.
> 

Thanks so much for the great debug method. 

I just locate the problem. For the same model, if I set the ```eps<=.01``` for ```dist.Gamma( eps, eps)```,  the graphical model I draw is,

![image](https://user-images.githubusercontent.com/24639671/143725019-799f133f-7b93-4634-bc78-9817cffd7a05.png)

But if I tune ```eps>=.05```, I get 

![image](https://user-images.githubusercontent.com/24639671/143725063-eb0b85d9-e6a3-4726-a52a-6825440a735a.png)

The second one is definitely what I want and that solves the problem. 
I do not have a good sense of GAMMA distribution, but is this what it should look like?

**Comment by fehiepsi** [comment]
I'm not sure what causes the issue - will take a closer look later next week. In the mean time, hopefully setting `numpyro.enable_x64()` at the beginning of your program (script/notebook) might solve the problem.

**Comment by EdwardRaff** [comment]
The issue with the graph changing for seemingly random reasons and having lots of arrows between almost all latents is something I've run into intermittently myself, but been unable to isolate / make occur consistently. Just as extra info that this isn't isolated.

**Comment by fehiepsi** [comment]
It seems that `dist.Gamma(1e-3, 1e-3)` has a very sharp peak at 0. I tried to get some samples in float64 and see that
```
import numpyro
import numpyro.distributions as dist
from jax import random
numpyro.enable_x64()

dist.Gamma(1e-3, 1e-3).sample(random.PRNGKey(0), (10,))
```
gives us very small values
```
DeviceArray([4.52041001e-144, 2.22507386e-305, 3.19962312e-041,
             6.90824557e-006, 2.22507386e-305, 2.22507386e-305,
             9.92220200e-155, 4.89574723e-208, 4.29673398e-131,
             2.22507386e-305], dtype=float64)
```
This leads to `nan` values when we try to use `grad` to inspect dependency in a model. The rendering code would be more robust when we have `ProvenanceArray` in #1217.

@fangzefunny Which algorithm does JAGS use? Does it use the Gibbs sampler? I can get some results with
```
    with npyro.plate("n_subj", obs['n_subj']):
        alpha = npyro.sample('alpha', dist.Normal(0, 1))
        a = npyro.sample('a', dist.Normal(0, 1))
        b = npyro.sample('b', dist.Normal(0, 1))
        alpha = (alpha * tau_alpha + mu_alpha)[..., None]
        a = jnp.clip(a * tau_a + mu_a, a_min=0., a_max=1.)[..., None]
        b = jnp.clip(b * tau_b + mu_b, a_min=0., a_max=1.)[..., None]
```
but not sure if it is what you want.
